### ISSUE TITLE
consider adding git hash/branch/tag into app

### ISSUE THREAD

**Issue by davidmaxwaterman** [Polymer/polymer-cli#835]
It seems like it would be easy enough for polymer-cli to add in a version number to an app, perhaps in the Polymer name space, and then make that available either directly or as an element(shock).

`polymer serve` could add it into the app dynamically, and `polymer build` could add it 'permanently'.

The use case is that I want to be sure what versions I have released to integration/production, and for users to be able to report bugs on specific versions they are using.

Having said that, with service worker caching files, a version number baked into each file is probably more suitable...but then that gets confusing with bundling.

TBH, I'm not sure what the ideal solution is, or if there is one, but it'd be good to be able to display a hash/etc inside some `(i)` dialog, at least - without the pain of moving to a custom build/gulpfile/etc.

**Comment by davidmaxwaterman** [comment]
Perhaps just adding the version number that's in the {polymer,bower,yarn,npm}.json files would suffice...or in addition.

**Comment by gdevacc12** [comment]
I agree with davidmaxwaterman. Having a time stamp from `polymer build` available to the app/element would be very useful, and/or an incremental version number. It would not need necessarily to be in all files. 

Disclaimer: I have never crafted gulp tasks (yet!) and although I have researched this I have not tested it (yet!). 

As I understand it: `polymer-cli` users `polymer-build` which uses `gulp`, so it should be possible to add a gulp task to build tasks, or as it seems most discussion goes using a custom polymer-build (not modifying polymer-cli. For example: [https://medium.com/collaborne-engineering/building-polymer-app-with-gulp-dee266d348df](url)
polymer-build is described here: [https://github.com/Polymer/polymer-build](url)
Either way  a simple replace a key string with a the relevant string generated at build time, for example using:
- `gulp-string-replace` [https://github.com/tomaszczechowski/gulp-string-replace](url) or
-  `gulp-replace-task` [https://github.com/outatime/gulp-replace-task](url)

Maybe this example would work: 
index.html content: 
`<html><body>some-unique-character- string-to-replace</body></html>`
additions to the gulp task doing the build:
`var replace = require('gulp-string-replace');
...
<in the gulp.src section>
.pipe(replace('some-unique-character- string-to-replace', (new Date()).toUTCString()))
`
In the usage section in the following link it shows how to augment the polymer build so this could be adapted. `https://github.com/Collaborne/gulp-polymer-build-utils/`

The just a matter of running gulp in the project root (apparently!)

**Comment by davidmaxwaterman** [comment]
Since you've given your solution, mine has been to abandon `polymer build` and use a gulpfile from somewhere (it has the polymer license header, so I guess it is a polymer one), and incorporated gulp-git and gulp-string-replace. I use the string '%TBCVERSION%' in my code where I want the version string to go (in html and js at the moment):

```
const replace = require('gulp-string-replace');
const git = require('gulp-git');
...
function build(version) {
...
          .pipe(gulpIf(/\.(js|html)$/, replace('%TBCVERSION%', version))) 
...
function version() {
  return new Promise(function(resolve, reject) {
    git.exec({args : 'describe --tags --dirty'}, function (err, stdout) {
      if (err) throw err;
      resolve(stdout.slice(0,-1));
    });
  });
}

gulp.task('build', function () {
  return version()
   .then((version) => build(version));
});
...
```

which produces a version based on my git tags, which is what I was already using (though not incorporating into the code).

Still, it would be much nicer if `polymer build` had this sort of thing built in.
